### DISCUSSION TITLE
new utility for generating bats-core tests

### DISCUSSION THREAD

**Discussion by tohara-PandoLogic** [bats-core/bats-core#698] [Show and tell]
Hi, I would like to highlight a new utility that can generate bats-core tests from bash file comments or plain text files. It also supports Jupyter notebooks using the Bash kernel
(see https://github.com/takluyver/bash_kernel). See

    BATSPP: Shell style tests using bats-core framework
    https://github.com/LimaBD/batspp

Here's is a simple example:

```
    $ cat hello-world.batspp
    # The hello world of Bats-pp tests

    $ echo 'Hello world' | wc -l                                                                                                           1
    1


    $ batspp hello-world.batspp 
    1..1
    ok 1 test of line 3
```
See the generated output at the end of this message.

The bats-pp utility was originally developed in support of testing my bash aliases and functions, but Bruno Limo has converted it into an independent project. In general, it supports testing command-line programs, checking either standard output or doing posthoc checks to produce testable output (e.g., file word counts).

The intention is similar in spirit to Python example-based tests from function docstrings. That is, the test specification shows the invocation and the expected output. Furthermore, the use of Jupiter notebooks allows for convenient maintenance of such tests. For example, if the programs change, the bats test files can be regenerated from the Jupyter notebook. For a variety of such notebooks, see https://github.com/tomasohara/shell-scripts/tree/aviyan-dev/tests.

Best,
Tom O'Hara
tomasohara@gmail.com

Generated bats-core test:
```

#!/usr/bin/env bats
#
# This test file was generated using Batspp
# https://github.com/LimaBD/batspp
#

# Constants
VERBOSE_DEBUG=""
TEMP_DIR="/home/tomohara/temp/tmp/batspp-15616"

# Setup function
# $1 -> test name
function run_setup () {
	test_folder=$(echo $TEMP_DIR/$1-$$)
	mkdir --parents "$test_folder"
	cd "$test_folder" || echo Warning: Unable to "cd $test_folder"
}

# Teardown function
function run_teardown () {
	: # Nothing here...
}

@test "test of line 3" {
	run_setup "test-of-line-3"

	# Assertion of line 3
	shopt -s expand_aliases
	print_debug "$(echo 'Hello world' | wc -l)" "$(echo -e '1')"
	[ "$(echo 'Hello world' | wc -l)" == "$(echo -e '1')" ]

	run_teardown
}

# This prints debug data when an assertion fail
# $1 -> actual value
# $2 -> expected value
function print_debug() {
	echo "=======  actual  ======="
	bash -c "echo \"$1\" $VERBOSE_DEBUG"
	echo "======= expected ======="
	bash -c "echo \"$2\" $VERBOSE_DEBUG"
	echo "========================"
}
```

**Comment by djmcgreal-cc** [comment]
Interesting! Via Jupyter was one of the way I wondered whether we should be presenting and testing documentation when I ended up going with https://github.com/bats-core/bats-core/discussions/694
